gammapy/maps/base.py

```diff
@@ -498,6 +499,119 @@ def upsample(self, factor, order=0, preserve_counts=True):
+
+        >>> # Create map
+        >>> m_wcs = Map.create(
+        >>>     coordsys='GAL',
+        >>>     map_type='wcs',
+        >>>     binsz=0.02,
+        >>>     skydir=skydir,
+        >>>     width=10.0,
+        >>>     axes=[energy_axis, time_axis],
+        >>>     unit='ct'
+        >>> )
+
+        >>> # Get image by coord tuple
+        >>> image = m_wcs.get_image_by_coord(('500 GeV', '1 h'))
+
+        >>> # Get image by coord dict with strings
+        >>> image = m_wcs.get_image_by_coord({'energy': '500 GeV', 'time': '1 h'})
+
+        >>> # Get image by coord dict with quantities
+        >>> image = m_wcs.get_image_by_coord({'energy': 0.5 * u.TeV, 'time': 1 * u.h})
+
+        Returns
+        -------
+        map_out : '~Map'
+            Map with spatial dimensions only.
+        """
+        if isinstance(coords, tuple):
+            axes_names = [_.name for _ in self.geom.axes]
+            coords = OrderedDict(zip(axes_names, coords))
+
+        coords_ = {'lon': np.nan, 'lat': np.nan}
+        coords_.update(coords)
+
+        for axis, value in zip(self.geom.axes, coords.values()):
+            coords_[axis.name] = u.Quantity(value).to(axis.unit).value
+
+        coords = MapCoord(coords_)
+        idx = self.geom.coord_to_idx(coords)
+        idx = idx[-len(self.geom.axes):]
+        return self.get_image_by_idx(idx, copy=copy)
+
+    def get_image_by_pix(self, pix, copy=True):
+        """Return spatial map at the given axis pixel coordinates
+
+        Parameters
+        ----------
+        pix : tuple
+            Tuple of pixel index arrays for each non-spatial dimension of the map.
+            Tuple should be ordered as (I_0, ..., I_n). Pixel indices can be
+            either float or integer type.
+
+        Returns
+        -------
+        map_out : '~Map'
+            Map with spatial dimensions only.
+        """
+        pix = (np.nan, ) * (self.data.ndim - len(self.geom.axes)) + pix
+        idx = self.geom.pix_to_idx(pix)
+        idx = idx[-len(self.geom.axes):]
+        return self.get_image_by_idx(idx, copy=copy)
+
+    def get_image_by_idx(self, idx, copy=True):
+        """Return spatial map at the given axis pixel indices.
+
+        Parameters
+        ----------
+        idx : tuple
+            Tuple of index arrays for each non spatial dimension of the map.
+            Tuple should be ordered as (I_0, ..., I_n).
+
+        Returns
+        -------
+        map_out : '~Map'
+            Map with spatial dimensions only.
+        """
+        if len(idx) != len(self.geom.axes):
+            raise ValueError("Tuple length must be equal to number of"
+                             " non spatial dimensions")
+        geom = self.geom.to_image()
+        map_out = self.__class__(geom=geom)
+        idx = (slice(None), ) * (self.data.ndim - len(self.geom.axes)) + idx
+        data_out = self.data.T[idx].copy() if copy else self.data.T[idx]
```

Comment:
This is complex.

Suggest to not duplicate the `self.data.T[idx]` code, but instead only have it once and the copy at the end:
```
if copy:
    data = data.copy()
```


Concerning the implemntation, can this be simplified?
- Can you avoid transpose and also squeeze?
- Is it possible to avoid the createion of the zeros map and assignment of new data? I.e. create data, geom and call Map constructor at the end. Should be ismpler and more efficient, no?

Reply:
Done